tests/devices/test_default_qubit_jax.py

```diff
@@ -259,14 +259,34 @@ def circuit(a):
 
         def cost(a):
             """A function of the device quantum state, as a function
-            of ijnput QNode parameters."""
+            of input QNode parameters."""
             res = jnp.abs(circuit(a)) ** 2
             return res[1] - res[0]
 
         grad = jax.grad(cost)(a)
         expected = jnp.sin(a)
         assert jnp.allclose(grad, expected, atol=tol, rtol=0)
 
+    def test_crot_differentiability(self, tol):
+        """Test that CRot can be differentiated."""
+        dev = qml.device("default.qubit.jax", wires=2)
+
+        @qml.qnode(dev, diff_method="backprop", interface="jax")
+        def circuit(a, b, c):
+            qml.PauliX(1)
+            qml.CRot(a, b, c, wires=[1, 0])
+            return qml.state()
+
+
+        def cost(a, b, c):
+            """A function of the device quantum state, as a function
+            of input QNode parameters."""
+            res = jnp.abs(circuit(a, b, c)) ** 2
+            return res[2] - res[3]
+
+        bg = jax.grad(cost, argnums=1)(0.54, 0.31, 1.22)
+        np.testing.assert_allclose(bg, -0.152529, atol=tol, rtol=tol)

```

Comment:
How did you calculate -0.152529? Just want to avoid a magic number in the test 😆

Reply:
By running the test :shame:

Reply:
You can adapt the autograd CRot test, which is analytic if you want:

```python
@pytest.mark.parametrize("theta", np.linspace(-2 * np.pi, np.pi, 7))
def test_CRot_gradient(self, theta, tol):
    """Tests that the automatic gradient of a arbitrary controlled Euler-angle-parameterized
    gate is correct."""
    dev = qml.device("default.qubit", wires=2)
    a, b, c = np.array([theta, theta ** 3, np.sqrt(2) * theta])

    @qml.qnode(dev, diff_method="parameter-shift")
    def circuit(a, b, c):
        qml.QubitStateVector(np.array([1., -1.]) / np.sqrt(2), wires=0)
        qml.CRot(a, b, c, wires=[0, 1])
        return qml.expval(qml.PauliX(0))

    res = circuit(a, b, c)
    expected = -np.cos(b / 2) * np.cos(0.5 * (a + c))
    assert np.allclose(res, expected, atol=tol, rtol=0)

    grad = qml.grad(circuit)(a, b, c)
    expected = np.array([[
        0.5 * np.cos(b / 2) * np.sin(0.5 * (a + c)),
        0.5 * np.sin(b / 2) * np.cos(0.5 * (a + c)),
        0.5 * np.cos(b / 2) * np.sin(0.5 * (a + c)),
    ]])
    assert np.allclose(grad, expected, atol=tol, rtol=0)
```

Reply:
also, whats wrong with magic numbers in tests? It seems just as likely the test math would be bugged as the actual code math. Hardcoded results minimizes that error chance.

Reply:
Changed it to be that test.

Reply:
> also, whats wrong with magic numbers in tests? It seems just as likely the test math would be bugged as the actual code math. Hardcoded results minimizes that error chance.

Less worried about the hardcoding, more about just making sure it's clear where the magic number came from for future developers.

We had issues earlier on that a lot of the gradient tests were comparing the computed gradient to the, uh, result of running the test. But the underlying gradient computation was buggy, and we didn't know about it for over a year (since the test always passed), and everyone just assumed the magic number was correct.

Reply:
Fair. Same thing would've happened here too.